File: Partition List/Partition_List.py
# 第一种思路，简单粗暴，使用两个列表，一个存放小节点，一个存放大节点
# 44ms 77.50%
class Solution:
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        smaller_stack, greater_stack = [], []
        while head is not None:
            if head.val < x:
                smaller_stack.append(head)
            else:
                greater_stack.append(head)
            head = head.next
        for index in range(len(smaller_stack) - 1):
            smaller_stack[index].next = smaller_stack[index + 1] 
        for index in range(len(greater_stack) - 1):
            greater_stack[index].next = greater_stack[index + 1]
        
        if smaller_stack:
            head = smaller_stack[0]
            if greater_stack:
                smaller_stack[-1].next = greater_stack[0]
                greater_stack[-1].next = None
            else:
                smaller_stack[-1].next = None
        else:
            if greater_stack:
                head = greater_stack[0]
                greater_stack[-1].next = None
            else:
                return None
        return head

File: Partition List/Partition_List.cpp
// Runtime: 8 ms, faster than 100.00% of C++ online submissions for Partition List.
// Memory Usage: 8.5 MB, less than 94.05% of C++ online submissions for Partition List.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution 
{
public:
    ListNode* partition(ListNode* head, int x) 
    {
        ListNode small(0), big(0);
        ListNode *smallPtr = &small, *bigPtr = &big;
        while (head)
        {
            if (head->val < x)
            {
                smallPtr->next = head;
                head = head->next;
                smallPtr = smallPtr->next;
                smallPtr->next = nullptr;
            }
            else
            {
                bigPtr->next = head;
                head = head->next;
                bigPtr = bigPtr->next;
                bigPtr->next = nullptr;
            }
        }
        smallPtr->next = big.next;
        return small.next;
    }
};